crates/turbopack-build/js/src/runtime.js

```diff
@@ -0,0 +1,359 @@
+/* eslint-disable @next/next/no-assign-module-variable */
+
+/** @typedef {import('../types').ModuleFactory} ModuleFactory */
+
+/** @typedef {import('../types').ChunkPath} ChunkPath */
+/** @typedef {import('../types').ModuleId} ModuleId */
+
+/** @typedef {import('../types').Module} Module */
+/** @typedef {import('../types').SourceInfo} SourceInfo */
+/** @typedef {import('../types').SourceType} SourceType */
+/** @typedef {import('../types').SourceType.Runtime} SourceTypeRuntime */
+/** @typedef {import('../types').SourceType.Parent} SourceTypeParent */
+/** @typedef {import('../types').Exports} Exports */
+/** @typedef {import('../types').EsmInteropNamespace} EsmInteropNamespace */
+
+const path = require("path");
+const relativePathToRuntimeRoot = path.relative(RUNTIME_PUBLIC_PATH, ".");
+const RUNTIME_ROOT = path.resolve(__filename, relativePathToRuntimeRoot);
+
+/** @type {Object.<ModuleId, ModuleFactory>} */
+const moduleFactories = { __proto__: null };
+/** @type {Object.<ModuleId, Module>} */
+const moduleCache = { __proto__: null };
+/**
+ * Module IDs that are instantiated as part of the runtime of a chunk.
+ *
+ * @type {Set<ModuleId>}
+ */
+const runtimeModules = new Set();
+/**
+ * Map from module ID to the chunks that contain this module.
+ *
+ * @type {Map<ModuleId, Set<ChunkPath>>}
+ */
+const moduleChunksMap = new Map();
+
+const hOP = Object.prototype.hasOwnProperty;
+const _process =
+  typeof process !== "undefined"
+    ? process
+    : {
+        env: {},
+        // Some modules rely on `process.browser` to execute browser-specific code.
+        // NOTE: `process.browser` is specific to Webpack.
+        browser: true,
+      };
+
+const toStringTag = typeof Symbol !== "undefined" && Symbol.toStringTag;
+
+/**
+ * @param {any} obj
+ * @param {PropertyKey} name
+ * @param {PropertyDescriptor & ThisType<any>} options
+ */
+function defineProp(obj, name, options) {
+  if (!hOP.call(obj, name)) Object.defineProperty(obj, name, options);
+}
+
+/**
+ * Adds the getters to the exports object
+ *
+ * @param {Exports} exports
+ * @param {Record<string, () => any>} getters
+ */
+function esm(exports, getters) {
```

Comment:
Yep, this could probably be shared as part of `turbopack-ecmascript`. The other solutions are:
* to create a crate for shared utils between `turbopack-dev` and `turbopack-build`.
* to consolidate both crates into a single crate `turbopack-ecmascript-runtime`.